modules/framework/cats/test/src-jvm/junit/JUnitRunnerTests.scala

```diff
@@ -0,0 +1,158 @@
+          TestStarted("foo(weaver.junit.Meta$Sharing)"),
+          TestFinished("foo(weaver.junit.Meta$Sharing)"),
+          TestSuiteFinished("weaver.junit.Meta$Sharing")
+        )
+        expect.same(notifications, expected)
+      }
+  }
+
+  def run(
+      blocker: BlockerCompat[IO],
+      suite: Class[_]): IO[List[Notification]] = for {
+    runner   <- IO(new WeaverRunner(suite))
+    queue    <- IO(scala.collection.mutable.Queue.empty[Notification])
+    notifier <- IO(new RunNotifier())
+    _        <- IO(notifier.addListener(new NotificationListener(queue)))
+    _        <- blocker.block(runner.run(notifier))
+  } yield queue.toList
+
+  def run(
+      blocker: BlockerCompat[IO],
+      suite: SimpleIOSuite): IO[List[Notification]] =
+    run(blocker, suite.getClass())
+
+  sealed trait Notification
+  case class TestSuiteStarted(name: String)             extends Notification
+  case class TestAssumptionFailure(failure: Failure)    extends Notification
+  case class TestFailure(name: String, message: String) extends Notification
+  case class TestFinished(name: String)                 extends Notification
+  case class TestIgnored(name: String)                  extends Notification
+  case class TestStarted(name: String)                  extends Notification
+  case class TestSuiteFinished(name: String)            extends Notification
+
+  class NotificationListener(
+      queue: scala.collection.mutable.Queue[Notification])
+      extends RunListener {
+    override def testSuiteStarted(description: Description): Unit =
+      queue += TestSuiteStarted(description.getDisplayName())
+    override def testAssumptionFailure(failure: Failure): Unit =
+      queue += TestAssumptionFailure(failure)
+    override def testFailure(failure: Failure): Unit =
+      queue += TestFailure(failure.getDescription.getDisplayName,
+                           failure.getMessage())
+    override def testFinished(description: Description): Unit =
+      queue += TestFinished(description.getDisplayName())
+    override def testIgnored(description: Description): Unit =
+      queue += TestIgnored(description.getDisplayName())
+    override def testStarted(description: Description): Unit =
+      queue += TestStarted(description.getDisplayName())
+    override def testSuiteFinished(description: Description): Unit =
+      queue += TestSuiteFinished(description.getDisplayName())
+  }
+
+}
+
+object Meta {
+
+  object MySuite extends SimpleIOSuite {
+
+    override def maxParallelism: Int = 1
+
+    pureTest("success") {
+      success
+    }
+
+    pureTest("failure") {
+      failure("oops")
+    }
+
+    test("ignore") {
+      ignore("just because")
+    }
+
+  }
+
+  object Only extends SimpleIOSuite {
+
+    override def maxParallelism: Int = 1
+
+    pureTest("only".only) {
+      success
+    }
+
+    pureTest("not only") {
+      failure("foo")
+    }
+
+  }
+
+  class Sharing(global: GlobalRead) extends SimpleIOSuite {
```

Comment:
Do we need tests here that actually use some global resources to make sure they're okay?

Reply:
bcd6d75